File: 526-beautiful-arrangement/526-beautiful-arrangement.py
class Solution:
    def countArrangement(self, N: int) -> int:
        states = set([i for i in range(1, N+1)])
        count = 0
        def helper(curr, i, states):
            if curr % i != 0 and i % curr != 0:
                return
            if len(states) == 0:
                nonlocal count
                count += 1
            for num in states:
                helper(num, i+1, states-set([num]))

        for i in states:
            helper(i, 1, states-set([i]))
        return count

File: 526-beautiful-arrangement/526-beautiful-arrangement.cpp
class Solution {
public:
    int countArrangement(int n) {
        vector<int> v(n);
        for(int i = 0; i < n; i++) v[i] = i + 1;
        int res = 0;
        help(v, res, 0);
        return res;
    }
    void help(vector<int>& v, int& res, int i) {
        if (i == v.size()) res++;
        else {
            for (int j = i; j < v.size(); j++) {
                swap(v[i], v[j]);
                if (v[i] % (i + 1) == 0 || (i + 1) % v[i] == 0) {
                    help(v, res, i + 1);
                }
                swap(v[i], v[j]);
            }
        }
    }
};